chapters/embeddings.qmd

```diff
@@ -1263,12 +1263,144 @@ There are a few notable patterns in these results:
+
+\begin{equation}
+\bar{x}^*_{jc}= \bar{x}_j + w_c\, s_j\, \delta^*_{jc}
+\end{equation}
+
+New features are added to the model based on the distance between $\boldsymbol{x}_0$ and $\boldsymbol{\bar{x}}^*_{c}$. The amount of shrinkage is best optimized using the tuning methods described in later chapters. There are several variations of this specific procedure. @wangImprovedCentroids describe several different approaches and @efron2009empirical demonstrates the connection to Bayesian methods. 
+
+```{r}
+#| label: ncs-comp
+#| results: hide
+
+set.seed(92)
+
+dat_1 <- MASS::mvrnorm(n = 200, c(1, -1), matrix(c(1, .5, .5, 1), 2, 2)) 
+dat_2 <- MASS::mvrnorm(n = 200, c(1, 3), matrix(c(1, -.75, -.75, 1), 2, 2))
+dat_3 <- MASS::mvrnorm(n = 200, c(-1, -1), matrix(c(1, 0, 0, 1), 2, 2))
+
+colnames(dat_1) <- colnames(dat_2) <- colnames(dat_3) <- c("x_1", "x_2")
+
+three_class_dat <- 
+  bind_rows(
+    dat_1 %>% as_tibble() %>% mutate(class = "class 1"),
+    dat_2 %>% as_tibble() %>% mutate(class = "class 2"),
+    dat_3 %>% as_tibble() %>% mutate(class = "class 3")
+  ) 
+
+dat <-
+  list(
+    x = t(as.matrix(three_class_dat[, 1:2])),
+    y = three_class_dat$class,
+    geneid = c("x_1", "x_2")
+  )
+res <- pamr::pamr.train(dat)
+cent_overall <- matrix(res$centroid.overall, ncol = 2)
+colnames(cent_overall) <- c("x_1", "x_2")
+cent_overall <- as_tibble(cent_overall)
+cent_class <- as_tibble(t(res$centroids)) %>% mutate(class = paste("class", 1:3))
+
+shrink <- function(threshold, object) {
+  res <- pamr::pamr.predict(object, threshold = threshold, type = "centroid")
+  out <- as_tibble(t(res))
+  out$class <- colnames(res)
+  out$threshold <- threshold
+  out
+}
+
+cent_shrunk <- map_dfr(res$threshold, shrink, object = res)
+```
+```{r}
+#| label: fig-nsc
+#| echo: false
+#| out-width: 80%
+#| fig-width: 6.25
+#| fig-height: 4
+#| fig-cap: An example of shrunken, class-specific centroids. Panel (a) shows data where there are three classes. Panel(b) demonstrates how, with increasing regularization, the centorids converge towards the global centroid (in black). 
+#| fig-alt: Nearest centroids
+
+three_class_points <- 
+  three_class_dat %>% 
+  ggplot(aes(x_1, x_2)) + 
+  geom_point(
+    aes(col = class, pch = class),
+    cex = 1,
+    alpha = 0.65,
+    show.legend = FALSE
+  ) +
+  labs(x = expression(x[1]), y = expression(x[2])) +
+  scale_color_brewer(palette = "Dark2") + 
+  scale_shape_manual(values = c(21, 22, 24)) +
+  ggtitle("(a)")
+
+three_class_path <- 
+  three_class_dat %>% 
+  ggplot(aes(x_1, x_2)) + 
+  geom_point(aes(col = class, pch = class), alpha = 0.27, pch = 1, cex = 1) +
+  geom_point(data = cent_class, aes(col = class, pch = class), cex = 3) +
+  geom_path(data = cent_shrunk, aes(col = class), show.legend = FALSE) +
+  geom_point(data = cent_overall, col = "black", cex = 3, pch = 16, show.legend = FALSE)  +
```

Comment:
```suggestion
  geom_point(data = cent_overall, col = "black", cex = 3, pch = 21, show.legend = FALSE)  +
```

Reply:
The purplish triangle (class 3) with the path made it look like the direction was pointing away from the centroid. I tweaked the figure to add arrows to better show the direction of the path to the centroid.  But please disregard if you prefer the original figure.